prow/tide/tide.go

```diff
@@ -756,6 +762,74 @@ func (c *Controller) takeAction(sp subpool, batchPending, successes, pendings, n
 	return Wait, nil, nil
 }
 
+// changedFilesAgent queries and caches the names of files changed by PRs.
+// Cache entries expire if they are not used during a sync loop.
+type changedFilesAgent struct {
+	ghc         githubClient
+	changeCache map[changeCacheKey][]string
+	// nextChangeCache caches file change info that is relevant this sync for use next sync.
+	// This becomes the new changeCache when prune() is called at the end of each sync.
+	nextChangeCache map[changeCacheKey][]string
+	sync.Mutex
+}
+
+type changeCacheKey struct {
+	org, repo string
+	number    int
+	sha       string
+}
+
+// prChanges gets the files changed by the PR, either from the cache or by
+// querying GitHub.
+func (c *changedFilesAgent) prChanges(pr *PullRequest) ([]string, error) {
+	cacheKey := changeCacheKey{
+		org:    string(pr.Repository.Owner.Login),
+		repo:   string(pr.Repository.Name),
+		number: int(pr.Number),
+		sha:    string(pr.HeadRefOID),
+	}
+
+	c.Lock()
```

Comment:
Could optimize with `RLock` here and `Lock` in `prune()`?

Reply:
this is only an optimization if you expect concurrent readers, RWMutex is more expensive